src/display/pattern_helper.js

```diff
@@ -322,63 +322,58 @@ var TilingPattern = (function TilingPatternClosure() {
 
       var x0 = bbox[0], y0 = bbox[1], x1 = bbox[2], y1 = bbox[3];
 
-      var topLeft = [x0, y0];
-      // we want the canvas to be as large as the step size
-      var botRight = [x0 + xstep, y0 + ystep];
+      // It is a waste of memory to use the size as specified by xstep or ystep
+      // if the bounding box is smaller.
+      xstep = Math.min(xstep, x1);
```

Comment:
... in case when Math.abs(XStep) is greater than cell width, we need to grow canvas, and in case when Math.abs(XStep) less than cell width we probably need to create a pattern that is painted on itself.

The canvas 2d context does not let us to set repeat distances so we probably need to do something clever, like paint in SVG, produce a canvas and "grow" it with SVG, disable "repeat" and repeat in fill/stroke. Also we need to apply optimizations for well-know paint patterns (e.g. painting simple image as using pattern) to replace with simple primitives.